network: wifi: check SSID when AP interfaces go up

sipraga opened the pull request:
When an AP goes up, the kernel may emit a netlink event indicating that
the interface has gained carrier. In that event, we should check if the
SSID has changed before attempting to reconfigure. Not doing so means
that the link->ssid member is not updated, leading to a potential
mismatch if some of the .network configurations match on SSID=.

There are however scenarios where the above heuristic is not enough.
Specifically, if the interface carrier state flip-flops within a short
enough interval, the internal throttling of netlink events inside the
kernel may suppress intermediate linkdown+linkup events (cf. Linux
net/core/link_watch.c). So there is no linkup event to react on.

To improve on the latter scenario, it is proposed to make newer kernels
emit an NL80211_CMD_START_AP multicast event when an AP goes up. This
event will not be dropped by link_watch. systemd-networkd can then react
to such events as well, and optionally reconfigure the link if the SSID
has changed. This will only work with newer kernels though.

sipraga commented:
This is a draft fix for https://github.com/systemd/systemd/issues/25384.

yuwata commented on src/network/networkd-link.c:
```diff
@@ -1498,6 +1499,11 @@ static int link_carrier_gained(Link *link) {
         link->previous_ssid = mfree(link->previous_ssid);
 
+        if (link->wlan_iftype == NL80211_IFTYPE_AP) {
+                link_get_wlan_interface(link);
```
Missing error handling.

yuwata commented on src/network/networkd-wifi.h:
```diff
@@ -7,3 +7,4 @@ typedef struct Manager Manager;
 
 int manager_genl_process_nl80211_config(sd_netlink *genl, sd_netlink_message *message, Manager *manager);
 int manager_genl_process_nl80211_mlme(sd_netlink *genl, sd_netlink_message *message, Manager *manager);
+int link_get_wlan_interface(Link *link);
```
Please add a typedef for `Link` in the above, like for `Manager`.

yuwata reviewed (changes requested):
Looks basically good to me. Several coding style requests.

sipraga commented:
Thanks @yuwata for checking. I still need to check something more on my end before I fix this up. I will update when ready.

sipraga commented:
Hi @yuwata

I have addressed the style comments you had.

In my testing, I found that the kernel's interface is not quite good enough to cover all scenarios. Specifically, if the AP undergoes rapid reconfigurations, it may lose and gain carrier a couple of times, but the kernel will silently drop linkdown/linkup events if the transition is short enough such that the time between a new linkup event and the last linkup event was less than 1 second.

This was very strange and finally I found my answer by reading the code in Linux related to link_watch.

Anyway, in order to improve on this situation, I would propose that the kernel emits an NL80211_CMD_START_AP on the multicast group whenever an AP goes up. That gives networkd something to react to. I have updated my commit to include handling for this scenario as well.

Please tell me what you think of this approach. I don't ask you to comment on whether the kernel change would be accepted (that is for Linux maintainers to decide), but rather your opinion as the maintainer of a consumer of this API (networkd).

If you find this agreeable, I will:

1. submit suitable changes to the linux-wireless mailing list to add the START_AP event
2. come back to this PR when the changes have landed in a stable kernel
3. then we can merge this

If you prefer, I can split the commit into two:

1. the original proposal I had (which covers the vast majority of realistic scenarios); this could be merged even today if you like
2. the extra code added to handle the new kernel event, which will make networkd fully robust.

Hope that makes sense. Thank you!

yuwata commented on src/network/networkd-wifi.c:
```diff
@@ -291,6 +291,39 @@ int manager_genl_process_nl80211_mlme(sd_netlink *genl, sd_netlink_message *mess
+                r = link_get_wlan_interface(link);
+                if (r < 0) {
+                        log_link_warning_errno(link, r, "Failed to update wireless LAN interface: %m");
+
```
Not a big matter, but please drop the empty line.

yuwata commented on src/network/networkd-link.c:
```diff
@@ -1498,6 +1499,12 @@ static int link_carrier_gained(Link *link) {
         force_reconfigure = link->previous_ssid && !streq_ptr(link->previous_ssid, link->ssid);
         link->previous_ssid = mfree(link->previous_ssid);
 
+        if (IN_SET(link->wlan_iftype, NL80211_IFTYPE_AP, NL80211_IFTYPE_P2P_GO)) {
```
Please add a brief comment why we need to actively update SSID or so.

yuwata reviewed (approved):
Looks mostly good. Several minor nitpicks.

Even if the kernel will provide better notification in the future, we need to still support the old kernel for a while. Hence, let's go with this logic.

sipraga commented:
> Looks mostly good. Several minor nitpicks.
> 
> Even if the kernel will provide better notification in the future, we need to still support the old kernel for a while. Hence, let's go with this logic.

Thanks! I will follow up later today and address your comments.

FYI, here is the kernel change I submitted (no feedback yet): https://lore.kernel.org/linux-wireless/20221209152836.1667196-1-alvin@pqrs.dk/

sipraga commented:
@yuwata address your comments, should be good to go now

yuwata commented:
LGTM. Thank you!
